File: apps/server/src/routes/achievement.ts
import { Router } from 'express';
import { prisma } from '../lib/prisma.js';
import { authMiddleware, type AuthRequest } from '../middleware/auth.js';
import { achievements } from '@sport-app/shared/achievements';

export const achievementRouter = Router();
achievementRouter.use(authMiddleware);

// GET /api/achievements — 成就墙（全部成就 + 解锁状态 + 进度）
achievementRouter.get('/', async (req: AuthRequest, res, next) => {
  try {
    const unlocked = await prisma.userAchievement.findMany({
      where: { userId: req.userId },
    });
    const unlockedMap = new Map(unlocked.map((a) => [a.achievementKey, a]));

    const all = achievements.map((def) => {
      const record = unlockedMap.get(def.key);
      return {
        key: def.key,
        name: def.name,
        description: def.description,
        icon: def.icon,
        category: def.category,
        rarity: def.rarity,
        unlocked: !!record,
        unlockedAt: record?.unlockedAt?.toISOString() || null,
      };
    });

    res.json({ success: true, data: all });
  } catch (err) {
    next(err);
  }
});

// GET /api/achievements/recent — 最近解锁的成就
achievementRouter.get('/recent', async (req: AuthRequest, res, next) => {
  try {
    const recent = await prisma.userAchievement.findMany({
      where: { userId: req.userId },
      orderBy: { unlockedAt: 'desc' },
      take: 10,
    });

    const withDetails = recent.map((record) => {
      const def = achievements.find((a) => a.key === record.achievementKey);
      return {
        key: record.achievementKey,
        name: def?.name || '未知成就',
        description: def?.description || '',
        icon: def?.icon || '',
        rarity: def?.rarity || 'common',
        unlockedAt: record.unlockedAt.toISOString(),
      };
    });

    res.json({ success: true, data: withDetails });
  } catch (err) {
    next(err);
  }
});

// GET /api/achievements/stats — 成就统计
achievementRouter.get('/stats', async (req: AuthRequest, res, next) => {
  try {
    const unlockedCount = await prisma.userAchievement.count({
      where: { userId: req.userId },
    });

    res.json({
      success: true,
      data: {
        unlocked: unlockedCount,
        total: achievements.length,
        rate: achievements.length > 0 ? Math.round((unlockedCount / achievements.length) * 100) : 0,
      },
    });
  } catch (err) {
    next(err);
  }
});

File: apps/server/src/routes/c/achievement.rs
use axum::{
    extract::State,
    routing::get,
    Json, Router,
};
use chrono::{DateTime, Utc};
use serde_json::json;

use crate::error::AppError;
use crate::middleware::auth::{AppState, AuthUser};
use crate::services::achievement::ACHIEVEMENTS;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/", get(achievement_wall))
        .route("/recent", get(recent_achievements))
        .route("/stats", get(achievement_stats))
}

// GET / — achievement wall
async fn achievement_wall(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<Json<serde_json::Value>, AppError> {
    let unlocked: Vec<(String, DateTime<Utc>)> = sqlx::query_as(
        r#"SELECT "achievementKey", "unlockedAt" FROM "UserAchievement" WHERE "userId" = $1"#,
    )
    .bind(&auth.user_id)
    .fetch_all(&state.pool)
    .await?;

    let wall: Vec<serde_json::Value> = ACHIEVEMENTS
        .iter()
        .map(|def| {
            let record = unlocked.iter().find(|(key, _)| key == def.key);
            json!({
                "key": def.key,
                "name": def.name,
                "description": def.description,
                "icon": def.icon,
                "category": def.category,
                "rarity": def.rarity,
                "unlocked": record.is_some(),
                "unlockedAt": record.map(|(_, t)| t.to_rfc3339()),
            })
        })
        .collect();

    Ok(Json(json!({ "success": true, "data": wall })))
}

// GET /recent — last 10 unlocked
async fn recent_achievements(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<Json<serde_json::Value>, AppError> {
    let rows: Vec<(String, DateTime<Utc>, Option<String>)> = sqlx::query_as(
        r#"SELECT "achievementKey", "unlockedAt", "unlockedByRun"
           FROM "UserAchievement" WHERE "userId" = $1 ORDER BY "unlockedAt" DESC LIMIT 10"#,
    )
    .bind(&auth.user_id)
    .fetch_all(&state.pool)
    .await?;

    let recent: Vec<serde_json::Value> = rows
        .into_iter()
        .filter_map(|(key, unlocked_at, unlocked_by_run)| {
            ACHIEVEMENTS.iter().find(|d| d.key == key).map(|def| {
                json!({
                    "key": def.key,
                    "name": def.name,
                    "description": def.description,
                    "icon": def.icon,
                    "rarity": def.rarity,
                    "unlockedAt": unlocked_at.to_rfc3339(),
                    "unlockedByRun": unlocked_by_run,
                })
            })
        })
        .collect();

    Ok(Json(json!({ "success": true, "data": recent })))
}

// GET /stats
async fn achievement_stats(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<Json<serde_json::Value>, AppError> {
    let unlocked: i64 = sqlx::query_scalar(
        r#"SELECT COUNT(*) FROM "UserAchievement" WHERE "userId" = $1"#,
    )
    .bind(&auth.user_id)
    .fetch_one(&state.pool)
    .await?;

    let total = ACHIEVEMENTS.len() as i64;
    let rate = if total > 0 { (unlocked as f64 / total as f64 * 100.0).round() as i64 } else { 0 };

    Ok(Json(json!({ "success": true, "data": { "unlocked": unlocked, "total": total, "rate": rate } })))
}
